pkg/util/purge/purge.go

```diff
@@ -0,0 +1,323 @@
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	return nil
+}
+
+// cleanNetworking lists subnets in vnets and unnassign security groups
+func (rc *ResourceCleaner) cleanNetworking(ctx context.Context, resourceGroup mgmtfeatures.ResourceGroup) error {
+	vnets, err := rc.vnetscli.List(ctx, *resourceGroup.Name)
+	if err != nil {
+		rc.log.Errorf("%s: %s", *resourceGroup.Name, err)
+		return err
+	}
+
+	for _, vnet := range vnets {
+		var changed bool
+
+		for i := range *vnet.Subnets {
+			if !rc.dryRun {
+				(*vnet.Subnets)[i].NetworkSecurityGroup = nil
+				changed = true
+			}
+		}
+
+		if changed {
+			rc.log.Printf("updating vnet %s/%s", *resourceGroup.Name, *vnet.Name)
+			if !rc.dryRun {
+				_, err := rc.vnetscli.CreateOrUpdate(ctx, *resourceGroup.Name, *vnet.Name, vnet)
+				if err != nil {
+					rc.log.Errorf(err.Error())
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// cleanPrivateLink lists and unassigns all private links. If they are assigned the deletoin will fail
+func (rc *ResourceCleaner) cleanPrivateLink(ctx context.Context, resourceGroup mgmtfeatures.ResourceGroup) error {
+	plss, err := rc.privatelinkservicescli.List(ctx, *resourceGroup.Name)
+	if err != nil {
+		return err
+	}
+	for _, pls := range plss {
+		for _, peconn := range *pls.PrivateEndpointConnections {
+			rc.log.Infof("deleting private endpoint connection %s/%s/%s", *resourceGroup.Name, *pls.Name, *peconn.Name)
+			if !rc.dryRun {
+				_, err := rc.privatelinkservicescli.DeletePrivateEndpointConnection(ctx, *resourceGroup.Name, *pls.Name, *peconn.Name)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+// CleanAAD does the cleaning of the Active Directory
+//     - clean unsused app registration
+//     - clean unsused role assignments
+func (rc *ResourceCleaner) CleanAAD(ctx context.Context) error {
+	err := rc.cleanApps(ctx)
+	if err != nil {
+		return err
+	}
+
+	return rc.cleanRoleAssignments(ctx)
+
+}
+
+// cleanRoleAssignments lists role assignments from the subscription and clear unused
+// the deletion work by:
+//     - checking whether the coresponding resource group exists (RS and RG have the same name)
+//     - check whether the assigned service principal exists
+//     - the role assignment belongs to one of RG: v4-westeurope, v4-eastus, v4-australiasoutheast
+func (rc *ResourceCleaner) cleanRoleAssignments(ctx context.Context) error {
+	roleAssignments, err := rc.roleassignmentcli.List(ctx, "")
+	if err != nil {
+		return err
+	}
+
+	// iterate over all roleAssignments and delete everything not associated with keep cluster
+	for roleAssignments.NotDone() {
+		pageResults := roleAssignments.Values()
+
+		for _, ra := range pageResults {
+			_, err := rc.resourcegroupscli.Get(ctx, *ra.Name)
+			if err != nil {
+				if detailedErr, ok := err.(autorest.DetailedError); ok &&
+					detailedErr.StatusCode == http.StatusNotFound {
+					continue
+				}
+				return err
+			}
```

Comment:
I don't understand this.  Isn't *ra.Name a UUID?

Reply:
It is, thanks for the pointer. I got confused by the portal representation.